### PR TITLE
Remove old rules when unbanning entities from BanLists.

### PR DESCRIPTION
**Pull request by Gnuxie**
Fixes https://github.com/matrix-org/mjolnir/issues/220

### CODE REVIEW

File: src/models/BanList.ts [hunk 1]
```diff
@@ -182,6 +182,38 @@ class BanList extends EventEmitter {
         return [...this.serverRules, ...this.userRules, ...this.roomRules];
     }
 
+    /**
+     * Remove all rules in the banList that would ban this entity by searching for rules that have legacy state types.
```
[the remark this answers is not in the document]
Comment (Gnuxie) [reply src/models/BanList.ts:186]: No, "match" doesn't mean a wildcard match. It just means remove all rules that have the same state key (as when we ban them) as there can be up to 3 different rules for the same entity (rarely at the same time though) depending on which state type was used for the rule when it was added.

File: src/models/BanList.ts [hunk 2]
```diff
@@ -182,6 +182,38 @@ class BanList extends EventEmitter {
...
+     * @param ruleType The normalized (most recent) type for this rule e.g. `RULE_USER`.
+     * @param entity The entity to unban from this list.
+     * @returns true if any rules were removed and the entity was unbanned, otherwise false because there were no rules.
+     */
+    public async unbanEntity(ruleType: string, entity: string): Promise<boolean> {
+        const stateKey = `rule:${entity}`;
+        const typesToCheck = (() => {
```
[the remark this answers is not in the document]
Comment (Gnuxie) [reply src/models/BanList.ts:193]: yeah, a `switch / case` is more noisy and I don't think they have an expressive version? I will change it though just so it doesn't throw people off

File: src/models/BanList.ts [hunk 3]
```diff
@@ -182,6 +182,38 @@ class BanList extends EventEmitter {
...
+            } else {
+                return [ruleType];
+            }
+        })();
+        // We can't cheat and check our state cache because we normalize the event types to the most recent version.
+        const typesToRemove = (await Promise.all(
+            typesToCheck.map(stateType => this.client.getRoomStateEvent(this.roomId, stateType, stateKey)
+                .then(_ => stateType) // We need the state type as getRoomState only returns the content, not the top level.
+                .catch(e => e.statusCode === 404 ? null : Promise.reject(e))))
+        ).filter(e => e); // remove nulls. I don't know why TS still thinks there can be nulls after this??
```
[the remark this answers is not in the document]
Comment (Gnuxie) [reply src/models/BanList.ts:209]: Yes, though to be clear I'm not talking about that null, TypeScript still complains that typesToRemove is `(string | null)[]` after the filter to remove them

File: src/models/BanList.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yoric) [reply src/models/BanList.ts]: Sadly, there's no expressive `switch`/`case` in JS :(

File: src/models/BanList.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yoric) [reply src/models/BanList.ts:210]: After the call to `filter(e => e)`? That's weird.

**Review by Yoric — APPROVED** [review]
Looks good, thanks!

File: src/models/BanList.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gnuxie) [reply src/models/BanList.ts:210]: Yeah, it is

**Pull request merged**